common/network/queues/concurrent_queue.c.inc

```diff
@@ -0,0 +1,110 @@
+  size_t CONCURRENT_QUEUE_OF_##TYPE##_SIZE(                \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    size_t size;                                           \
+    if (self == NULL) return 0;                            \
+    lock_handle_lock(&self->lock);                         \
+    size = self->size;                                     \
+    lock_handle_unlock(&self->lock);                       \
+    return size;                                           \
+  }
+
+#define DEFINE_FUNCTION_PUSH_CONCURRENT_QUEUE_OF(TYPE)                      \
+  concurrent_queue_status_t CONCURRENT_QUEUE_OF_##TYPE##_PUSH(              \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE const data) { \
+    concurrent_queue_node_of_##TYPE##_t *node;                              \
+    if (self == NULL) return CONCURRENT_QUEUE_NULL_SELF;                    \
+    node = calloc(1, sizeof(*node));                                        \
+    if (node == NULL) return CONCURRENT_QUEUE_ALLOC_FAILED;                 \
+    node->data = data;                                                      \
+    lock_handle_lock(&self->lock);                                          \
+    if (self->front == NULL) {                                              \
+      self->front = node;                                                   \
+      self->back = node;                                                    \
+    } else {                                                                \
+      self->back->next = node;                                              \
+      self->back = node;                                                    \
+    }                                                                       \
+    self->size++;                                                           \
+    lock_handle_unlock(&self->lock);                                        \
+    cond_handle_signal(&self->cond);                                        \
+    return CONCURRENT_QUEUE_SUCCESS;                                        \
+  }
+
+#define DEFINE_FUNCTION_POP_CONCURRENT_QUEUE_OF(TYPE)                        \
+  concurrent_queue_status_t CONCURRENT_QUEUE_OF_##TYPE##_POP(                \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE *const data) { \
+    if (self == NULL) return CONCURRENT_QUEUE_NULL_SELF;                     \
+    if (data == NULL) return CONCURRENT_QUEUE_NULL_DATA;                     \
+    lock_handle_lock(&self->lock);                                           \
+    while (self->size == 0) {                                                \
+      cond_handle_wait(&self->cond, &self->lock);                            \
+    }                                                                        \
+    concurrent_queue_node_of_##TYPE##_t *tmp = self->front;                  \
+    *data = tmp->data;                                                       \
```

Comment:
this won't work? you'd need to do a memcpy of tmp->data to data?

Reply:
tmp->data is not a pointer so it's a simple copy